sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2099,9 +2099,14 @@ class Analyzer(override val catalogManager: CatalogManager)
         ResolvedFunc(Identifier.of(funcIdent.database.toArray, funcIdent.funcName))
 
       case q: LogicalPlan =>
+        val isGroupingIdAllowed = q match {
+          case _: Aggregate | _: GroupingSets | _: Filter | _: Sort => true
+          case _ => false
+        }
         q transformExpressions {
           case u if !u.childrenResolved => u // Skip until children are resolved.
-          case u: UnresolvedAttribute if resolver(u.name, VirtualColumn.hiveGroupingIdName) =>
```

Comment:
This PR can make some error messages change? If yes, could you add some tests for the case? e.g.,

```
scala> sql("""select `grouping__id` from values (1, 1)  t(a, b)""").explain(true)
org.apache.spark.sql.AnalysisException: grouping_id() can only be used with GroupingSets/Cube/Rollup;;
!Project [grouping_id() AS grouping__id#22]
+- SubqueryAlias t
   +- LocalRelation [a#23, b#24]
```